core/src/main/java/org/apache/accumulo/core/spi/crypto/AESCryptoService.java

```diff
@@ -89,6 +89,17 @@ public class AESCryptoService implements CryptoService {
 
   private static final FileEncrypter DISABLED = new NoFileEncrypter();
 
+  private static final ThreadLocal<Cipher> KEY_WRAP_CIPHER = new ThreadLocal<Cipher>() {
+    @Override
+    protected Cipher initialValue() {
+      try {
+        return Cipher.getInstance(KEY_WRAP_TRANSFORM);
```

Comment:
Calls to Cipher.getInstance are expensive and include a synchronization point deep down the call stack. Ciphers are *not* thread-safe, but they are reusable. This Cipher is only used for Key wrapping and unwrapping. I adding the `synchronized` modifier to the two methods that are using this object.

Reply:
Why use a ThreadLocal? Couldn't you just get the instance once when the crypto service is first created?

Reply:
because the Cipher is not thread-safe.

Reply:
Yeah but I don't see how using a `ThreadLocal` is going to help with that. Is this to avoid using any schronization?

Reply:
I think I see what you are saying, we would need one Cipher in a ThreadLocal for wrapping and another for unwrapping.

Reply:
Created separate ThreadLocal's in 1e289c7

Reply:
I'm not sure this is the best approach. I thought you created a `ThreadLocal` to avoid extra calls to Cipher.getInstance? According to the ThreadLocal javadoc... "each thread that accesses one (via its get or set method) has its own, independently initialized copy of the variable." That sounds like it won't help us.

I will take a look at this today with the goal to create less calls to Cipher.getInstance() and keep it thread safe.

Reply:
It does reduce the number of calls. It reduces it to one per-thread for the lifetime of the VM instead of one per invocation of unwrapKey/wrapKey.